coderd/httpmw/actor.go

```diff
@@ -0,0 +1,70 @@
+package httpmw
+
+import (
+	"net/http"
+
+	"github.com/coder/coder/coderd/httpapi"
+	"github.com/coder/coder/coderd/rbac"
+	"github.com/coder/coder/codersdk"
+)
+
+// RequireAPIKeyOrExternalProxyAuth is middleware that should be inserted after
+// optional ExtractAPIKey and ExtractExternalProxy middlewares to ensure one of
+// the two authentication methods is provided.
+//
+// If both are provided, an error is returned to avoid misuse.
+func RequireAPIKeyOrExternalProxyAuth() func(http.Handler) http.Handler {
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			_, hasAPIKey := APIKeyOptional(r)
+			_, hasExternalProxy := ExternalProxyOptional(r)
+
+			if hasAPIKey && hasExternalProxy {
+				httpapi.Write(r.Context(), w, http.StatusBadRequest, codersdk.Response{
+					Message: "API key and external proxy authentication provided, but only one is allowed",
+				})
+				return
+			}
```

Comment:
This is not used to save code duplication but to avoid complicating the existing endpoints that moons need to be able to access. E.g. moons need to be able to access every workspace agent to be able to connect to it, but instead of duplicating the endpoint we're using the same endpoint and just allowing the actor to be a moon actor.

I'm not too fond of this way of doing it, but because of how inextensible our authz/authn scheme is this seems like the best option to me other than duplicating the whole endpoint.

Reply:
We did not remove this, we still have this "Only one can be set".

We only want the client to set the user auth header or the workspace proxy header. They can set both headers, but it is not clear what to do in that situation, so it is best to throw an error. If we do not throw an error, then we are silently choosing 1 auth over the other, and could lead to an annoying debugging session.

---

We accept both because a workspace proxy needs to access the same routes as a regular user.